docs/root/configuration/http/http_filters/tcp_post_filter.rst

```diff
@@ -0,0 +1,132 @@
+.. _config_http_filters_tcp_post:
+
+TCP Post
+===============
+
+The TcpPost filter is used by the upstream proxy to coordinate with downstream proxy that proxies
+TCP streams over HTTP/2 POST requests. 
+
+Configuration
+-------------
+
+* This filter should be configured with the name *envoy.filters.http.tcp_post*.
+
+
+It converts POST requests back to CONNECT to trigger the regular TCP decapping. Therefore, it's
+typically used with another localhost listener that does the regular TCP decapping.
```

Comment:
What's the reason that this filter only converts to CONNECT but not TCP directly?

Reply:
The actual TCP decap is done by this regular envoy CONNECT support, not by this filter:

https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/http/upgrades#connect-support

Reply:
I know TCP decap can be done by regular CONNECT support, but having an extra cluster/listener isn't really efficient because all traffic have to go through localhost twice. I'm asking what is the reason not do that within the filter (might including some refactor to reuse that part).

Reply:
Gotcha, @alyssawilk please chime in.

My thought is that I am not sure if this feature essential enough to change the core of regular TCP decamping, which is triggered by CONNECT.

Reply:
I thought by adding a config option we were going to configure which methods trigger TCP decap.  CONNECT by default but others if configured.  WDYT?

Reply:
SGTM. Let me give it a try.

Reply:
PTAL at the new API in api/envoy/config/route/v3/route_components.proto.

The binary change is also surprisingly simple :)

I also added example configs for encaping and terminating.

If this looks reasonable to you, I still move forward to add tests, docs and clean up the http filter.

Thanks!

cc @mattklein123

Reply:
I'm a strong +1 on just adding a config option to the route. I think it's much simpler and I think it's reasonable to have it be in the core. @lizan @alyssawilk?

/wait

Reply:
I think configuring the method in the upgrade configs is simpler and I'm fine with it being in core if you are :-)

Reply:
May I ask what is the "core" ?

IIUC, the API I added is in "route_components" - "upgrade_config" - "connect_connfig".

Reply:
oh sorry, actually in the default TCP/HTTP code which currently special cases CONNECT, rather than in an filter (which may or may not be included in a given Envoy build)

Reply:
I am still a bit confused. I wonder:

1) is route_components the right place to put the API? if not, what is the right place?
2) if yes, is connect_config the right place?
3) what does "special cases CONNECT" mean? I looked at the codes, it seems like router is the only place that chooses a tcp_stream for decaping. I don't see any places, such as http filter, http_connection_manager, would trigger the tcp decaping logic.

Reply:
Per offline discussion, the api added to router is preferred.